tools/proto2jsonschema/postgen.js

```diff
@@ -84,46 +79,41 @@ console.log(`Fixing paths in dir: ${ROOT}`);
     cleanSchema(resultJSON, true);
 
     /**
-     * Simplify the $ref tags. Used in some languages like Java.
+     * Simplify and fix the $ref tags. This string is used for the name of fields.
      *
      * This couldn't be done in the previous step, because the $ref was simply wrong in that step.
      * Now the $refs are correct (but long).
      *
      * We need to change the $ref name and definition.
      *
      * We do so by:
-     * - traversing the json, getting all refs
-     * - getting the simplified replacement $ref without the prefix
-     * - replaceAll ref ids (fixes "definition" and corresponding "$ref" tags)
+     * - Looking at the JSON schema "definitions".
```

Comment:
This seems prone to introducing naming collisions. For example, if there are two references, one to "google.protobuf.Value" and one to "google.events.cloud.firestore.v1.Value", won't those both end up as "Value"?

Reply:
I thought about this. TLDR it's literally impossible for a collision unless the proto has a collision `message`.

JSON schema field naming doesn't work like proto naming.

JSON schema has a set of "definitions" and "$ref". All schema definitions are directly in the JSON Schema (at least in our design), no external definitions. The definitions are literally the messages in the proto, so they are "Document" and "Value" in Firestore:

https://github.com/googleapis/google-cloudevents/blob/master/proto/google/events/cloud/firestore/v1/data.proto

`google.protobuf.Value` isn't a JSON Schema definition, and should never be. Only the messages within the data proto can be a JSON Schema definition (if re-used).

---

This is not a new design, it's just implementing the existing design after updating the protoc plugin.

Reply:
So how are references to other messages handled? Do they end up as "unnamed" schema? (I'm guessing that's the case, looking at the Cloud Audit LogEntryData JSON schema.)

(I'm still not entirely sure about this design given that it means two uses of the same "external" proto presumably aren't represented by a common class within the generated library, but that's a separate matter.)

Reply:
References to other messages are handled via `$ref` fields. I don't know what you're referring to w.r.t. unnamed messages.

I think a new GitHub issue is the best way to address the concern. I'm not sure if you're talking about cross-schema references or some other issue.

Reply:
I've filed https://github.com/googleapis/google-cloudevents/issues/180, describing the concern and how I propose to test it. I'll get to that when I've got schema generation working locally.